### PR TITLE
[GlobalISel] Change MatchTable entries to 1 byte each

### PR DESCRIPTION
**Pull request by Pierre-vh**
See https://discourse.llvm.org/t/rfc-make-globalisel-match-table-entries-1-byte-instead-of-8/75411

This helps reduce llc's binary size, at the cost of some added complexity to the MatchTable machinery.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h
Comment (wangpc-pp) [inline llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:1389]: This casting is weird.

File: llvm/utils/TableGen/GlobalISelMatchTable.cpp
Comment (wangpc-pp) [inline llvm/utils/TableGen/GlobalISelMatchTable.cpp:51]: Why not just put these prints in `emitEncodingMacrosDef` and `emitEncodingMacrosUndef` to  `emitMatchTable`? So that we can see the whole structure of generated code.

File: llvm/utils/TableGen/GlobalISelMatchTable.cpp
Comment (wangpc-pp) [inline llvm/utils/TableGen/GlobalISelMatchTable.cpp:267]: The maximum output bytes would be 10 I think.

File: llvm/utils/TableGen/GlobalISelMatchTable.cpp
Comment (wangpc-pp) [inline llvm/utils/TableGen/GlobalISelMatchTable.cpp:333]: Can this be commited seperately so that we can see the bytes reduction?
It can be `CurrentSize * 8` bytes without your changes.

**Review by wangpc-pp — COMMENTED** [review]
The code is LGTM in general, but the changes are large and I need more eyes.

File: llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h
Comment (Pierre-vh) [reply llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h:462]: Why? It's more verbose
Coding style says it's fine if it's casting between integral types:
https://llvm.org/docs/CodingStandards.html#prefer-c-style-casts

> When casting between integral types (including enums that are not strongly- typed), functional-style casts are permitted as an alternative to static_cast.

File: llvm/utils/TableGen/GlobalISelMatchTable.cpp
Comment (Pierre-vh) [reply llvm/utils/TableGen/GlobalISelMatchTable.cpp:333]: Is it alright if I just git push that straight away or do I need to do another PR?

File: llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h
Comment (Pierre-vh) [reply llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h]: Indeed but it's needed, I've added a `readS8` wrapper around it with a comment explaining why we use that.

File: llvm/utils/TableGen/GlobalISelMatchTable.cpp
Comment (wangpc-pp) [reply llvm/utils/TableGen/GlobalISelMatchTable.cpp:333]: Please create another PR for it and add some test coverages (you can just add some CHECKs to check the table size).

**Review by wangpc-pp — APPROVED** [review]
LGTM!
I'd like to see this great work landed but please wait for one more approval from other reviewers to see if they have some comments.

File: llvm/utils/TableGen/GlobalISelMatchTable.cpp
Comment (Pierre-vh) [reply llvm/utils/TableGen/GlobalISelMatchTable.cpp:45]: "need", no, it's just a style choice. I like putting constants that are reused multiple times into a variable

**Comment by Pierre-vh** [comment]
@arsenm / @aemerson is this good to land?

**Review by qcolombet — APPROVED** [review]
Given we already talked about compile time in a previous thread and things were looking good, this LGTM

**Comment by aemerson** [comment]
Please hold off for now, I'm still trying to collect data.

**Comment by aemerson** [comment]
So I built clang with release mode and no assertions, with thinlto. I used fc791b612723 as the baseline and:
```
commit 44f6d94c58b8073432f52b368258414f9afd0ae4 (compact-gisel-match-table)
Author: pvanhout <pierre.vanhoutryve@amd.com>
Date:   Fri Dec 8 08:56:30 2023 +0100

    Simplify getEncodedEmitStr
```
as the test commit.

Building CTMark in the test suite with `-j1` and 10 runs, here's the data for -Os, with `compare.py` using `--merge-average`.
```
Program                                       compile_time
                                              before        after  diff
tramp3d-v4/tramp3d-v4                           5.37         5.42  0.8%
mafft/pairlocalalign                            2.51         2.53  0.5%
kimwitu++/kc                                    6.45         6.47  0.4%
ClamAV/clamscan                                 4.83         4.84  0.2%
sqlite3/sqlite3                                 2.35         2.35  0.1%
SPASS/SPASS                                     4.26         4.26  0.0%
lencod/lencod                                   4.39         4.39 -0.0%
Bullet/bullet                                  10.28        10.27 -0.1%
7zip/7zip-benchmark                            13.91        13.90 -0.1%
consumer-typeset/consumer-typeset               3.39         3.38 -0.3%
                           Geomean difference                      0.2%
```
So it seems there's a 0.2% geomean regression but I'm not fully convinced this isn't noise.

I decided to dig further by trying to run llc on a sqlite3 optimized bitcode file, to test just codegen performance and try to reduce the effect of noise from the rest of compilation/file IO. Doing this and measuring CPU cycles I saw a 0.5% mean regression.

I then looked at your PR here: https://github.com/llvm/llvm-project/pull/74823 and measuring this resulted in a no statistically significant change in CTMark overall compilation times, however I did see a 0.8% reduction in average cycles for the llc test, compared to the baseline. So it seems that PR brings a needed tiny perf improvement.

Overall, given that this change is a predecessor to that PR and that it does give a huge size improvement, I think this is ok to go.

**Comment by wangpc-pp** [comment]
> I then looked at your PR here: #74823 and measuring this resulted in a no statistically significant change in CTMark overall compilation times vs baseline (so no regression anymore), however I did see a 0.8% reduction in average cycles for the llc test, compared to the baseline. So it seems that PR brings a needed tiny perf improvement.

I think the measurement result of #74823 may be noise too, because that PR reduces less code size than this PR. 
In a word, the measurement result shows no regression here. So, I think we are OK to go further?

**Comment by aemerson** [comment]
> > I then looked at your PR here: #74823 and measuring this resulted in a no statistically significant change in CTMark overall compilation times vs baseline (so no regression anymore), however I did see a 0.8% reduction in average cycles for the llc test, compared to the baseline. So it seems that PR brings a needed tiny perf improvement.
> 
> 
> 
> I think the measurement result of #74823 may be noise too, because that PR reduces less code size than this PR. 
> 
> In a word, the measurement result shows no regression here. So, I think we are OK to go further?

I think that's possible perhaps but unlikely. These measurements were from 40 runs each, directly reading PMU counters. The root cause for the improvement I don't know yet, but yes agreed that either way it's worth the binary size reduction.

**Comment by Pierre-vh** [comment]
Thanks a lot for the extensive testing @aemerson :)

**Comment by vitalybuka** [comment]
Breaks https://lab.llvm.org/buildbot/#/builders/5/builds/39231/steps/9/logs/stdio
Could you please fix or revert

**Comment by vitalybuka** [comment]
> Breaks https://lab.llvm.org/buildbot/#/builders/5/builds/39231/steps/9/logs/stdio Could you please fix or revert

Bot should be fixed with eabf7ec3f3d4688ea0f9da24038462362837d7ff

BTW. Does it need to take into account `llvm::endian`?

**Comment by Pierre-vh** [comment]
> > Breaks https://lab.llvm.org/buildbot/#/builders/5/builds/39231/steps/9/logs/stdio Could you please fix or revert
> 
> Bot should be fixed with [eabf7ec](https://github.com/llvm/llvm-project/commit/eabf7ec3f3d4688ea0f9da24038462362837d7ff)
> 
> BTW. Does it need to take into account `llvm::endian`?

I don't think it needs to, the host machine should read the value as big/little endian naturally. The encoding part in TableGen is endian-aware and encodes it depending on big/little endian though

**Pull request merged**
